File: pkg/pathfmt/pathfmt.go
package pathfmt

import (
	"bytes"
	"path"
	"regexp"
	"strings"
	"text/template"
)

type Data struct {
	Event      string
	Zone       string
	MatchName  string
	Order      int
	RedSchool  string
	RedName    string
	BlueSchool string
	BlueName   string
	RoundNo    int
	Role       string
}

func Render(matchNameTemplate, pathTemplate string, data Data) (string, error) {
	data, err := withMatchName(matchNameTemplate, data)
	if err != nil {
		return "", err
	}
	out, err := execute(pathTemplate, data)
	if err != nil {
		return "", err
	}
	return sanitizePath(out), nil
}

func RenderMatchDir(matchNameTemplate, matchDirTemplate string, data Data) (string, error) {
	data, err := withMatchName(matchNameTemplate, data)
	if err != nil {
		return "", err
	}
	out, err := execute(matchDirTemplate, data)
	if err != nil {
		return "", err
	}
	return sanitizePath(out), nil
}

func withMatchName(matchNameTemplate string, data Data) (Data, error) {
	name, err := execute(matchNameTemplate, data)
	if err != nil {
		return data, err
	}
	data.MatchName = name
	return data, nil
}

func execute(tpl string, data Data) (string, error) {
	t, err := template.New("path").Parse(tpl)
	if err != nil {
		return "", err
	}
	var buf bytes.Buffer
	if err := t.Execute(&buf, data); err != nil {
		return "", err
	}
	return buf.String(), nil
}

var badSegmentChars = regexp.MustCompile(`[<>:"\\|?*\x00-\x1f]`)

func sanitizePath(p string) string {
	parts := strings.FieldsFunc(p, func(r rune) bool { return r == '/' || r == '\\' })
	for i, part := range parts {
		part = strings.TrimSpace(part)
		part = badSegmentChars.ReplaceAllString(part, "_")
		part = strings.Trim(part, ". ")
		if part == "" {
			part = "_"
		}
		parts[i] = part
	}
	return path.Join(parts...)
}

File: danmu-job/src/pathfmt.ts
import path from "node:path";
import type { RecordConf } from "./config.js";

export type PathData = {
  Event: string;
  Zone: string;
  MatchName?: string;
  Order: number;
  RedSchool: string;
  RedName: string;
  BlueSchool: string;
  BlueName: string;
  RoundNo: number;
  Role: string;
};

export function renderMatchDir(conf: Required<RecordConf>, data: PathData): string {
  const withName = {
    ...data,
    MatchName: executeTemplate(conf.MatchNameTemplate, data),
  };
  return sanitizePath(executeTemplate(conf.MatchDirTemplate, withName));
}

export function resolveUnderBase(base: string, rel: string): string {
  if (!rel) {
    return path.normalize(base);
  }
  if (path.isAbsolute(rel) || rel.startsWith("/")) {
    return path.normalize(rel);
  }
  return path.join(base, rel.split("/").join(path.sep));
}

function executeTemplate(template: string, data: PathData): string {
  return template.replace(/\{\{\s*\.([A-Za-z0-9_]+)\s*\}\}/g, (_, key: keyof PathData) => {
    const value = data[key];
    return value === undefined || value === null ? "" : String(value);
  });
}

export function sanitizePath(input: string): string {
  const parts = input
    .split(/[\\/]/)
    .map((part) => {
      let out = part.trim().replace(/[<>:"\\|?*\x00-\x1f]/g, "_");
      out = out.replace(/^[. ]+|[. ]+$/g, "");
      return out || "_";
    })
    .filter((part) => part.length > 0);
  return parts.join("/");
}
